src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs

```diff
@@ -105,7 +105,12 @@ internal DataContractJsonSerializer(Type type,
         internal ISerializationSurrogateProvider? SerializationSurrogateProvider
         {
             get { return _serializationSurrogateProvider; }
-            set { _serializationSurrogateProvider = value; }
+            set
+            {
+                _serializationSurrogateProvider = value;
+                // Reset _rootContract when surrogate provider changes so it gets recalculated
+                _rootContract = null;
+            }
```

Comment:
This "cache invalidation" is not necessary and it is not done in the big brother DataContractSerializer. Please undo this change.